File: packages/firebase_common/lib/src/firestore_documents/host_location.flutterfire_gen.dart
// coverage:ignore-file
// ignore_for_file: type=lint
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target, unnecessary_question_mark

part of 'host_location.dart';

class ReadHostLocation {
  const ReadHostLocation({
    required this.hostLocationId,
    required this.path,
    required this.hostId,
    required this.address,
    required this.geo,
    required this.createdAt,
    required this.updatedAt,
  });

  final String hostLocationId;

  final String path;

  final String hostId;

  final String address;

  final Geo geo;

  final DateTime? createdAt;

  final DateTime? updatedAt;

  factory ReadHostLocation._fromJson(Map<String, dynamic> json) {
    return ReadHostLocation(
      hostLocationId: json['hostLocationId'] as String,
      path: json['path'] as String,
      hostId: json['hostId'] as String,
      address: json['address'] as String? ?? '',
      geo: _geoConverter.fromJson(json['geo'] as Map<String, dynamic>),
      createdAt: (json['createdAt'] as Timestamp?)?.toDate(),
      updatedAt: (json['updatedAt'] as Timestamp?)?.toDate(),
    );
  }

  factory ReadHostLocation.fromDocumentSnapshot(DocumentSnapshot ds) {
    final data = ds.data()! as Map<String, dynamic>;
    return ReadHostLocation._fromJson(<String, dynamic>{
      ...data,
      'hostLocationId': ds.id,
      'path': ds.reference.path,
    });
  }
}

class CreateHostLocation {
  const CreateHostLocation({
    required this.hostId,
    required this.address,
    required this.geo,
  });

  final String hostId;
  final String address;
  final Geo geo;

  Map<String, dynamic> toJson() {
    return {
      'hostId': hostId,
      'address': address,
      'geo': _geoConverter.toJson(geo),
      'createdAt': FieldValue.serverTimestamp(),
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }
}

class UpdateHostLocation {
  const UpdateHostLocation({
    this.hostId,
    this.address,
    this.geo,
    this.createdAt,
  });

  final String? hostId;
  final String? address;
  final Geo? geo;
  final DateTime? createdAt;

  Map<String, dynamic> toJson() {
    return {
      if (hostId != null) 'hostId': hostId,
      if (address != null) 'address': address,
      if (geo != null) 'geo': _geoConverter.toJson(geo!),
      if (createdAt != null) 'createdAt': createdAt,
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }
}

class DeleteHostLocation {}

/// Provides a reference to the hostLocations collection for reading.
final readHostLocationCollectionReference = FirebaseFirestore.instance
    .collection('hostLocations')
    .withConverter<ReadHostLocation>(
      fromFirestore: (ds, _) => ReadHostLocation.fromDocumentSnapshot(ds),
      toFirestore: (_, __) => throw UnimplementedError(),
    );

/// Provides a reference to a hostLocation document for reading.
DocumentReference<ReadHostLocation> readHostLocationDocumentReference({
  required String hostLocationId,
}) =>
    readHostLocationCollectionReference.doc(hostLocationId);

/// Provides a reference to the hostLocations collection for creating.
final createHostLocationCollectionReference = FirebaseFirestore.instance
    .collection('hostLocations')
    .withConverter<CreateHostLocation>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (obj, _) => obj.toJson(),
    );

/// Provides a reference to a hostLocation document for creating.
DocumentReference<CreateHostLocation> createHostLocationDocumentReference({
  required String hostLocationId,
}) =>
    createHostLocationCollectionReference.doc(hostLocationId);

/// Provides a reference to the hostLocations collection for updating.
final updateHostLocationCollectionReference = FirebaseFirestore.instance
    .collection('hostLocations')
    .withConverter<UpdateHostLocation>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (obj, _) => obj.toJson(),
    );

/// Provides a reference to a hostLocation document for updating.
DocumentReference<UpdateHostLocation> updateHostLocationDocumentReference({
  required String hostLocationId,
}) =>
    updateHostLocationCollectionReference.doc(hostLocationId);

/// Provides a reference to the hostLocations collection for deleting.
final deleteHostLocationCollectionReference = FirebaseFirestore.instance
    .collection('hostLocations')
    .withConverter<DeleteHostLocation>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (_, __) => throw UnimplementedError(),
    );

/// Provides a reference to a hostLocation document for deleting.
DocumentReference<DeleteHostLocation> deleteHostLocationDocumentReference({
  required String hostLocationId,
}) =>
    deleteHostLocationCollectionReference.doc(hostLocationId);

/// Manages queries against the hostLocations collection.
class HostLocationQuery {
  /// Fetches [ReadHostLocation] documents.
  Future<List<ReadHostLocation>> fetchDocuments({
    GetOptions? options,
    Query<ReadHostLocation>? Function(Query<ReadHostLocation> query)?
        queryBuilder,
    int Function(ReadHostLocation lhs, ReadHostLocation rhs)? compare,
  }) async {
    Query<ReadHostLocation> query = readHostLocationCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    final qs = await query.get(options);
    final result = qs.docs.map((qds) => qds.data()).toList();
    if (compare != null) {
      result.sort(compare);
    }
    return result;
  }

  /// Subscribes [HostLocation] documents.
  Stream<List<ReadHostLocation>> subscribeDocuments({
    Query<ReadHostLocation>? Function(Query<ReadHostLocation> query)?
        queryBuilder,
    int Function(ReadHostLocation lhs, ReadHostLocation rhs)? compare,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    Query<ReadHostLocation> query = readHostLocationCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    var streamQs =
        query.snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamQs = streamQs.where((qs) => !qs.metadata.hasPendingWrites);
    }
    return streamQs.map((qs) {
      final result = qs.docs.map((qds) => qds.data()).toList();
      if (compare != null) {
        result.sort(compare);
      }
      return result;
    });
  }

  /// Fetches a specific [ReadHostLocation] document.
  Future<ReadHostLocation?> fetchDocument({
    required String hostLocationId,
    GetOptions? options,
  }) async {
    final ds = await readHostLocationDocumentReference(
      hostLocationId: hostLocationId,
    ).get(options);
    return ds.data();
  }

  /// Subscribes a specific [HostLocation] document.
  Stream<ReadHostLocation?> subscribeDocument({
    required String hostLocationId,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    var streamDs = readHostLocationDocumentReference(
      hostLocationId: hostLocationId,
    ).snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamDs = streamDs.where((ds) => !ds.metadata.hasPendingWrites);
    }
    return streamDs.map((ds) => ds.data());
  }

  /// Adds a [HostLocation] document.
  Future<DocumentReference<CreateHostLocation>> add({
    required CreateHostLocation createHostLocation,
  }) =>
      createHostLocationCollectionReference.add(createHostLocation);

  /// Sets a [HostLocation] document.
  Future<void> set({
    required String hostLocationId,
    required CreateHostLocation createHostLocation,
    SetOptions? options,
  }) =>
      createHostLocationDocumentReference(
        hostLocationId: hostLocationId,
      ).set(createHostLocation, options);

  /// Updates a specific [HostLocation] document.
  Future<void> update({
    required String hostLocationId,
    required UpdateHostLocation updateHostLocation,
  }) =>
      updateHostLocationDocumentReference(
        hostLocationId: hostLocationId,
      ).update(updateHostLocation.toJson());

  /// Deletes a specific [HostLocation] document.
  Future<void> delete({
    required String hostLocationId,
  }) =>
      deleteHostLocationDocumentReference(
        hostLocationId: hostLocationId,
      ).delete();
}

File: functions/src/firestore-documents/hostLocation.flutterfireGen.ts
import * as admin from 'firebase-admin'
import {
    DocumentData,
    DocumentReference,
    DocumentSnapshot,
    FieldValue,
    GeoPoint,
    Query,
    QueryDocumentSnapshot,
    QuerySnapshot,
    SetOptions,
    Timestamp,
    WriteResult
} from 'firebase-admin/firestore'

export class Geo {
    constructor({
        geohash,
        geopoint
    }: {
        geohash: string
        geopoint: GeoPoint
    }) {
        this.geohash = geohash
        this.geopoint = geopoint
    }

    readonly geohash: string

    readonly geopoint: GeoPoint
}

export class ReadHostLocation {
    constructor({
        hostLocationId,
        path,
        hostId,
        address,
        geo,
        createdAt,
        updatedAt
    }: {
        hostLocationId: string
        path: string
        hostId: string
        address: string
        geo: Geo
        createdAt?: Date
        updatedAt?: Date
    }) {
        this.hostLocationId = hostLocationId
        this.path = path
        this.hostId = hostId
        this.address = address
        this.geo = geo
        this.createdAt = createdAt
        this.updatedAt = updatedAt
    }

    readonly hostLocationId: string

    readonly path: string

    readonly hostId: string

    readonly address: string

    readonly geo: Geo

    readonly createdAt?: Date

    readonly updatedAt?: Date

    private static geoConverterFromJson(json: Record<string, unknown>): Geo {
        const geohash = json[`geohash`] as string
        const geopoint = json[`geopoint`] as GeoPoint
        return new Geo({ geohash, geopoint })
    }

    private static fromJson(json: Record<string, unknown>): ReadHostLocation {
        return new ReadHostLocation({
            hostLocationId: json[`hostLocationId`] as string,
            path: json[`path`] as string,
            hostId: json[`hostId`] as string,
            address: (json[`address`] as string | undefined) ?? ``,
            geo: ReadHostLocation.geoConverterFromJson(
                json[`geo`] as Record<string, unknown>
            ),
            createdAt: (json[`createdAt`] as Timestamp | undefined)?.toDate(),
            updatedAt: (json[`updatedAt`] as Timestamp | undefined)?.toDate()
        })
    }

    static fromDocumentSnapshot(ds: DocumentSnapshot): ReadHostLocation {
        const data = ds.data()!
        const cleanedData: Record<string, unknown> = {}
        for (const [key, value] of Object.entries(data)) {
            cleanedData[key] = value === null ? undefined : value
        }
        return ReadHostLocation.fromJson({
            ...cleanedData,
            hostLocationId: ds.id,
            path: ds.ref.path
        })
    }
}

export class CreateHostLocation {
    constructor({
        hostId,
        address,
        geo
    }: {
        hostId: string
        address: string
        geo: Geo
    }) {
        this.hostId = hostId
        this.address = address
        this.geo = geo
    }

    readonly hostId: string

    readonly address: string

    readonly geo: Geo

    private static geoConverterToJson(geo: Geo): Record<string, unknown> {
        return {
            geohash: geo.geohash,
            geopoint: geo.geopoint
        }
    }

    toJson(): Record<string, unknown> {
        return {
            hostId: this.hostId,
            address: this.address,
            geo: CreateHostLocation.geoConverterToJson(this.geo),
            createdAt: FieldValue.serverTimestamp(),
            updatedAt: FieldValue.serverTimestamp()
        }
    }
}

export class UpdateHostLocation {
    constructor({
        hostId,
        address,
        geo,
        createdAt
    }: {
        hostId?: string
        address?: string
        geo?: Geo
        createdAt?: Date
    }) {
        this.hostId = hostId
        this.address = address
        this.geo = geo
        this.createdAt = createdAt
    }

    readonly hostId?: string

    readonly address?: string

    readonly geo?: Geo

    readonly createdAt?: Date

    private static geoConverterToJson(geo: Geo): Record<string, unknown> {
        return {
            geohash: geo.geohash,
            geopoint: geo.geopoint
        }
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    toJson(): Record<string, any> {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const json: Record<string, any> = {}
        if (this.hostId != undefined) {
            json[`hostId`] = this.hostId
        }
        if (this.address != undefined) {
            json[`address`] = this.address
        }
        if (this.geo != undefined) {
            json[`geo`] = UpdateHostLocation.geoConverterToJson(this.geo)
        }
        if (this.createdAt != undefined) {
            json[`createdAt`] = this.createdAt
        }
        json[`updatedAt`] = FieldValue.serverTimestamp()
        return json
    }
}

export class DeleteHostLocation {}

/**
 * A Cloud Firestore object which ignores `undefined` properties.
 */
const db = admin.firestore()
db.settings({ ignoreUndefinedProperties: true })

/**
 * Provides a reference to the hostLocations collection for reading.
 */
export const readHostLocationCollectionReference = db
    .collection(`hostLocations`)
    .withConverter<ReadHostLocation>({
        fromFirestore: (ds: DocumentSnapshot): ReadHostLocation => {
            return ReadHostLocation.fromDocumentSnapshot(ds)
        },
        toFirestore: () => {
            throw new Error(
                `toFirestore is not implemented for ReadHostLocation`
            )
        }
    })

/**
 * Provides a reference to a hostLocation document for reading.
 * @param hostLocationId - The ID of the hostLocation document to read.
 */
export const readHostLocationDocumentReference = ({
    hostLocationId
}: {
    hostLocationId: string
}): DocumentReference<ReadHostLocation> =>
    readHostLocationCollectionReference.doc(hostLocationId)

/**
 * Provides a reference to the hostLocations collection for creating.
 */
export const createHostLocationCollectionReference = db
    .collection(`hostLocations`)
    .withConverter<CreateHostLocation>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for CreateHostLocation`
            )
        },
        toFirestore: (obj: CreateHostLocation): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a hostLocation document for creating.
 * @param hostLocationId - The ID of the hostLocation document to read.
 */
export const createHostLocationDocumentReference = ({
    hostLocationId
}: {
    hostLocationId: string
}): DocumentReference<CreateHostLocation> =>
    createHostLocationCollectionReference.doc(hostLocationId)

/**
 * Provides a reference to the hostLocations collection for updating.
 */
export const updateHostLocationCollectionReference = db
    .collection(`hostLocations`)
    .withConverter<UpdateHostLocation>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for UpdateHostLocation`
            )
        },
        toFirestore: (obj: UpdateHostLocation): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a hostLocation document for updating.
 * @param hostLocationId - The ID of the hostLocation document to read.
 */
export const updateHostLocationDocumentReference = ({
    hostLocationId
}: {
    hostLocationId: string
}): DocumentReference<UpdateHostLocation> =>
    updateHostLocationCollectionReference.doc(hostLocationId)

/**
 * Provides a reference to the hostLocations collection for deleting.
 */
export const deleteHostLocationCollectionReference = db
    .collection(`hostLocations`)
    .withConverter<DeleteHostLocation>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for DeleteHostLocation`
            )
        },
        toFirestore: (): DocumentData => {
            throw new Error(
                `toFirestore is not implemented for DeleteHostLocation`
            )
        }
    })

/**
 * Provides a reference to a hostLocation document for deleting.
 * @param hostLocationId - The ID of the hostLocation document to read.
 */
export const deleteHostLocationDocumentReference = ({
    hostLocationId
}: {
    hostLocationId: string
}): DocumentReference<DeleteHostLocation> =>
    deleteHostLocationCollectionReference.doc(hostLocationId)

/**
 * Manages queries against the hostLocations collection.
 */
export class HostLocationQuery {
    /**
     * Fetches hostLocation documents.
     * @param queryBuilder - Function to modify the query.
     * @param compare - Function to sort the results.
     */
    async fetchDocuments({
        queryBuilder,
        compare
    }: {
        queryBuilder?: (
            query: Query<ReadHostLocation>
        ) => Query<ReadHostLocation>
        compare?: (lhs: ReadHostLocation, rhs: ReadHostLocation) => number
    }): Promise<ReadHostLocation[]> {
        let query: Query<ReadHostLocation> = readHostLocationCollectionReference
        if (queryBuilder != undefined) {
            query = queryBuilder(query)
        }
        const qs: QuerySnapshot<ReadHostLocation> = await query.get()
        let result = qs.docs.map(
            (qds: QueryDocumentSnapshot<ReadHostLocation>) => qds.data()
        )
        if (compare != undefined) {
            result = result.sort(compare)
        }
        return result
    }

    /**
     * Fetches a specific hostLocation document.
     * @param hostLocationId - The ID of the hostLocation document to fetch.
     */
    async fetchDocument({
        hostLocationId
    }: {
        hostLocationId: string
    }): Promise<ReadHostLocation | undefined> {
        const ds = await readHostLocationDocumentReference({
            hostLocationId
        }).get()
        return ds.data()
    }

    /**
     * Adds a hostLocation document.
     * @param createHostLocation - The hostLocation details to add.
     */
    async add({
        createHostLocation
    }: {
        createHostLocation: CreateHostLocation
    }): Promise<DocumentReference<CreateHostLocation>> {
        return createHostLocationCollectionReference.add(createHostLocation)
    }

    /**
     * Sets a hostLocation document.
     * @param hostLocationId - The ID of the hostLocation document to set.
     * @param createHostLocation - The hostLocation details to set.
     * @param options - Options for the set operation.
     */
    async set({
        hostLocationId,
        createHostLocation,
        options
    }: {
        hostLocationId: string
        createHostLocation: CreateHostLocation
        options?: SetOptions
    }): Promise<WriteResult> {
        if (options == undefined) {
            return createHostLocationDocumentReference({
                hostLocationId
            }).set(createHostLocation)
        } else {
            return createHostLocationDocumentReference({
                hostLocationId
            }).set(createHostLocation, options)
        }
    }

    /**
     * Updates a specific hostLocation document.
     * @param hostLocationId - The ID of the hostLocation document to update.
     * @param updateHostLocation - The details for updating the hostLocation.
     */
    async update({
        hostLocationId,
        updateHostLocation
    }: {
        hostLocationId: string
        updateHostLocation: UpdateHostLocation
    }): Promise<WriteResult> {
        return updateHostLocationDocumentReference({
            hostLocationId
        }).update(updateHostLocation.toJson())
    }

    /**
     * Deletes a specific hostLocation document.
     * @param hostLocationId - The ID of the hostLocation document to delete.
     */
    async delete({
        hostLocationId
    }: {
        hostLocationId: string
    }): Promise<WriteResult> {
        return deleteHostLocationDocumentReference({
            hostLocationId
        }).delete()
    }
}
